client/X11/xf_gdi.c

```diff
@@ -970,96 +970,79 @@ static BOOL xf_gdi_surface_update_frame(xfContext* xfc, UINT16 tx, UINT16 ty,
-			rect.width = cmd->width;
-			rect.height = cmd->height;
-			XSetFunction(xfc->display, xfc->gc, GXcopy);
-			XSetFillStyle(xfc->display, xfc->gc, FillSolid);
-			XSetClipRectangles(xfc->display, xfc->gc, cmd->destLeft, cmd->destTop,
-			                   &rect, 1, YXBanded);
+			                         gdi->height, NULL))
+				goto fail;
 
-			/* Invalidate the updated region */
-			if (!xf_gdi_surface_update_frame(xfc, rect.x, rect.y,
-			                                 rect.width, rect.height))
-				ret = FALSE;
-
-			XSetClipMask(xfc->display, xfc->gc, None);
+			ret = xf_gdi_update_screen(xfc, cmd, gdi->primary_buffer);
 			break;
 
 		case RDP_CODEC_ID_NSCODEC:
 			if (!nsc_process_message(context->codecs->nsc, cmd->bpp, cmd->width,
 			                         cmd->height,
 			                         cmd->bitmapData, cmd->bitmapDataLength,
 			                         gdi->primary_buffer, gdi->dstFormat, 0, 0, 0, cmd->width, cmd->height))
-			{
-				xf_unlock_x11(xfc, FALSE);
-				return FALSE;
-			}
+				goto fail;
 
-			XSetFunction(xfc->display, xfc->gc, GXcopy);
-			XSetFillStyle(xfc->display, xfc->gc, FillSolid);
-			pDstData = gdi->primary_buffer;
-			image = XCreateImage(xfc->display, xfc->visual, xfc->depth, ZPixmap, 0,
-			                     (char*) pDstData, cmd->width, cmd->height, xfc->scanline_pad, 0);
-			XPutImage(xfc->display, xfc->primary, xfc->gc, image, 0, 0,
-			          cmd->destLeft, cmd->destTop, cmd->width, cmd->height);
-			XFree(image);
-			ret = xf_gdi_surface_update_frame(xfc, cmd->destLeft, cmd->destTop, cmd->width,
-			                                  cmd->height);
-			XSetClipMask(xfc->display, xfc->gc, None);
+			ret = xf_gdi_update_screen(xfc, cmd, gdi->primary_buffer);
 			break;
 
 		case RDP_CODEC_ID_NONE:
-			XSetFunction(xfc->display, xfc->gc, GXcopy);
-			XSetFillStyle(xfc->display, xfc->gc, FillSolid);
 			pSrcData = cmd->bitmapData;
-			pDstData = gdi->primary_buffer;
-			freerdp_image_copy(pDstData, gdi->dstFormat, 0, 0, 0,
-			                   cmd->width, cmd->height, pSrcData,
-			                   PIXEL_FORMAT_BGRX32_VF, 0, 0, 0, &xfc->context.gdi->palette);
-			image = XCreateImage(xfc->display, xfc->visual, xfc->depth, ZPixmap, 0,
-			                     (char*) pDstData, cmd->width, cmd->height, xfc->scanline_pad, 0);
-			XPutImage(xfc->display, xfc->primary, xfc->gc, image, 0, 0,
-			          cmd->destLeft, cmd->destTop,
-			          cmd->width, cmd->height);
-			XFree(image);
-			ret = xf_gdi_surface_update_frame(xfc, cmd->destLeft, cmd->destTop, cmd->width,
-			                                  cmd->height);
-			XSetClipMask(xfc->display, xfc->gc, None);
+
+			if (!freerdp_image_copy(gdi->primary_buffer, gdi->dstFormat, 0, 0, 0,
+			                        cmd->width, cmd->height, pSrcData,
+			                        PIXEL_FORMAT_BGRX32_VF, 0, 0, 0, &xfc->context.gdi->palette))
+				goto fail;
+
+			ret = xf_gdi_update_screen(xfc, cmd, gdi->primary_buffer);
 			break;
 
```

Comment:
maybe all the calls:
_ret = xf_gdi_update_screen(xfc, cmd, gdi->primary_buffer);_ 
could be mutualized at the end of the switch() (excluding the default case of course)